### PR TITLE
Don't ask Hive for statistics for hidden columns

### PR DESCRIPTION
**Pull request by alandpost**
Hidden columns are internal to Presto.  Presto should not ask Hive for
statistics on them.  The code path for getting stats on an unpartitioned
table (inside MetastoreHiveStatisticsProvider.getPartitionsStatistics())
was asking Hive for stats on hidden columns, which for some
implementations of Hive (such as CDH 5.11.0) could throw an error.  This
patch hoists the check up to outside the HiveStatisticsProvider.

This solves #8380.

### CODE REVIEW

**Comment by alandpost** [comment]
This came from a Teradata support case.

Note that there is no testing in this patch.  What would the best approach be?

**Comment by cawallin** [comment]
You should be able to create a test in `AbstractTestHiveClient` -- via `presto/hive-hadoop2/bin/run_on_docker.sh`, this is an integration test that runs against Hadoop in docker.

This solves https://github.com/prestodb/presto/issues/8380.

File: presto-hive/src/main/java/com/facebook/presto/hive/statistics/HiveStatisticsProvider.java [hunk 1]
```diff
@@ -23,6 +23,9 @@
 import java.util.List;
 import java.util.Map;
 
+/**
+ * @param tableColumns must be hive columns, not hidden (presto-internal) columns
```
Comment (electrum) [inline presto-hive/src/main/java/com/facebook/presto/hive/statistics/HiveStatisticsProvider.java]: Nit: capitalize Presto

**Comment by losipiuk** [comment]
> You should be able to create a test in AbstractTestHiveClient -- via presto/hive-hadoop2/bin/run_on_docker.sh, this is an integration test > that runs against Hadoop in docker.

It could be correct approach, though I am not very familiar with those tests. Especially I am not sure how (and if) is it possible to populate instance of hive Metastore they start with stats for tables. Normally you would do that with `ANALYZE` on Hive but it requires some `Hive` runtime (e.g. M/R) and if I understand correctly those tests are lacking it. Maybe one can set the stats manualy using some thrift call in test setup()?

So far we only have product tests for integration of Hive and Presto concerning stats in `TestHiveTableStatistics`. But it is far from optimal.

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java [hunk 2]
```diff
@@ -396,7 +396,9 @@ public TableStatistics getTableStatistics(ConnectorSession session, ConnectorTab
             return EMPTY_STATISTICS;
         }
         List<HivePartition> hivePartitions = partitionManager.getPartitions(metastore, tableHandle, constraint).getPartitions();
-        Map<String, ColumnHandle> tableColumns = getColumnHandles(session, tableHandle);
+        Map<String, ColumnHandle> tableColumns = Maps.filterValues(
```
Comment (electrum) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java]: `filterValues` returns a view which can have unexpected performance characteristics. For this reason, we avoid using Guava methods that return collection views.

This could be wrapped with `ImmutableMap.copyOf()`, but we can easily use streams instead:
```java
Map<String, ColumnHandle> tableColumns = getColumnHandles(session, tableHandle)
        .entrySet().stream()
        .filter(entry -> ((HiveColumnHandle) entry.getValue()).isHidden())
        .collect(toImmutableMap(Map.Entry::getKey, Map.Entry::getValue));

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java [hunk 3]
```diff
@@ -396,7 +396,9 @@ public TableStatistics getTableStatistics(ConnectorSession session, ConnectorTab
...
+                getColumnHandles(session, tableHandle),
+                columnHandle -> !getColumnMetadata(session, tableHandle, columnHandle).isHidden());
```
Comment (electrum) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java]: No need to convert to `ColumnMetadata` -- just cast to `HiveColumnHandle`

**Comment by electrum** [comment]
@losipiuk The test tables for `AbstractTestHiveClient` are setup using Hive via `create-test.sql` in `run_on_docker.sh`

(This is just FYI. I have no opinion on where these tests should go.)

**Comment by losipiuk** [comment]
> @losipiuk The test tables for AbstractTestHiveClient are setup using Hive via create-test.sql in run_on_docker.sh

Thanks @electrum.

@alandpost it looks like it should be possible to call Hive's `ANALYZE` to compute statistics from the `create-test.sql` file.
At this point M/R services should be running in docker (based on contents of `run_on_docker.sh`). 
It feels to me that it would be beneficial to have coverage of how we access Hive statistics from metastore at this level.
Product tests cover much more layers and only validate what we currently have in SPI and what is exposed via `SHOW STATS` statement.

**Comment by alandpost** [comment]
@electrum I made the changes you requested.

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java [hunk 4]
```diff
@@ -396,7 +397,10 @@ public TableStatistics getTableStatistics(ConnectorSession session, ConnectorTab
             return EMPTY_STATISTICS;
         }
         List<HivePartition> hivePartitions = partitionManager.getPartitions(metastore, tableHandle, constraint).getPartitions();
-        Map<String, ColumnHandle> tableColumns = getColumnHandles(session, tableHandle);
+        Map<String, ColumnHandle> tableColumns = getColumnHandles(session, tableHandle)
+                .entrySet().stream()
```
Comment (nezihyigitbasi) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java]: `.stream()` on a new line

Comment (alandpost) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java]: @nezihyigitbasi The code in question was dictated to me by @electrum -- perhaps the two of you can have an offline discussion about this.

Comment (kokosing) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java]: The best option to avoid such situations would be to add a checkstyle rule which would enforce the proper formatting, then we could avoid such not-the-first-priority comments.

**Comment by alandpost** [comment]
I added a HiveMetadata-level basic test for partitioned and unpartitioned table stats.

File: presto-spi/src/main/java/com/facebook/presto/spi/statistics/ColumnStatistics.java [hunk 5]
```diff
@@ -45,6 +47,39 @@ public Estimate getDistinctValuesCount()
         return distinctValuesCount;
     }
 
+    @Override
```
Comment (losipiuk) [inline presto-spi/src/main/java/com/facebook/presto/spi/statistics/ColumnStatistics.java]: Please extract separate commit with adding toString/hashCode/equals to Stats SPI.

File: presto-hive/src/test/sql/create-test.sql [hunk 6]
```diff
@@ -232,3 +232,12 @@ ALTER TABLE presto_test_partition_schema_change REPLACE COLUMNS (t_data DOUBLE);
 
 INSERT OVERWRITE TABLE presto_test_partition_schema_change_non_canonical PARTITION (t_boolean='0')
 SELECT 'test' FROM presto_test_sequence LIMIT 100;
+
+ANALYZE TABLE presto_test_sequence COMPUTE STATISTICS;
```
Comment (losipiuk) [inline presto-hive/src/test/sql/create-test.sql]: Do we need stats for all of those tables. 
How much does it change test time?

File: presto-spi/src/main/java/com/facebook/presto/spi/statistics/ColumnStatistics.java [hunk 7]
```diff
@@ -45,6 +47,39 @@ public Estimate getDistinctValuesCount()
...
+    public boolean equals(Object o)
+    {
+        if (this == o) {
+            return true;
+        }
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+        ColumnStatistics columnStatistics = (ColumnStatistics) o;
+        return getDataSize().equals(columnStatistics.getDataSize()) &&
```
Comment (losipiuk) [inline presto-spi/src/main/java/com/facebook/presto/spi/statistics/ColumnStatistics.java]: Is that generated. Generaly (from what I see) Presto uses `Objects.equals` in `equals()` methods. Please regenrate using that.

File: presto-hive/src/test/sql/create-test.sql (hunk 6 above)
Comment (losipiuk) [inline presto-hive/src/test/sql/create-test.sql]: To have column statistics you have to run
`ANALYZE TABLE .... COMPUTE STATISTICS FOR COLUMNS`.

File: presto-hive/src/test/java/com/facebook/presto/hive/AbstractTestHiveClient.java [hunk 8]
```diff
@@ -1058,6 +1061,74 @@ public void testGetTableSchemaException()
     }
 
     @Test
+    public void testGetTableStatsBucketedStringInt()
```
Comment (losipiuk) [inline presto-hive/src/test/java/com/facebook/presto/hive/AbstractTestHiveClient.java]: It would be better to have those tests as separate PR as they are not really related to hidden-column stuff.
I extracted first commit here https://github.com/Teradata/presto/commit/7fda962ce57f1ff842f202eb0f1bb6a3d13143ab.
And I will merge it as Travis passes.

Comment (losipiuk) [inline presto-hive/src/test/java/com/facebook/presto/hive/AbstractTestHiveClient.java]: Merged first commit. Please rebase this PR and change title.

**Comment by alandpost** [comment]
Test moved to https://github.com/prestodb/presto/pull/8670
